File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int answer = -1;

        int low = 0, high = nums.length-1;
        int mid;
        while(low<=high) {
            mid = low + (high-low)/2;

            /* target 값이면 그대로 반환 */
            if(nums[mid] == target) return mid;

            /* asc인 구간을 찾아 이진탐색 */
            if(nums[mid] >= nums[low]) {
                if(nums[low] <= target && nums[mid] >= target) high = mid-1;
                else low = mid+1;
            }else {
                if(nums[mid] <= target && nums[high] >= target) low = mid+1;
                else high = mid-1;
            }
        }
        return answer;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (l + r) // 2
            if target == nums[m]:
                return m

            # middle is in left portion
            if nums[l] <= nums[m]:
                if target < nums[l] or nums[m] < target:
                    l = m + 1
                else:
                    r = m - 1
            else:
                if target < nums[m] or nums[r] < target:
                    r = m - 1
                else:
                    l = m + 1

        return -1
